File: reDiscover/TGPlaylistViewController.m
//
//  TGPlaylistViewController.m
//  Proto3
//
//  Created by Teo Sartori on 01/07/2013.
//  Copyright (c) 2013 Teo Sartori. All rights reserved.
//

#import "TGPlaylistViewController.h"
#import "TGPlaylistCellView.h"
#import "TGPlaylist.h"
#import "TGSongPool.h"

@interface TGPlaylistViewController () <TGPlaylistDelegate>
@end

@implementation TGPlaylistViewController

-(void)awakeFromNib {
    if (playlist == NULL) {
        playlist = [[TGPlaylist alloc] init];
        [playlist setDelegate:self];
        
        // Set this controller as the playlist table view's delegate and data source.
        [_playlistTableView setDelegate:self];
        [_playlistTableView setDataSource:self];
       
    }
}

-(void)addSongToPlaylist:(id<SongIDProtocol>)aSongID {
    [playlist addSong:aSongID atIndex:0];
    [_playlistTableView reloadData];
}


-(void)removeSongFromPlaylist:(id)aSong {
    [playlist removeSong:aSong];
    [_playlistTableView reloadData];
}

- (id)getNextSongIDToPlay {
    return [playlist getNextSongIDToPlay];
}

- (void)storePlaylistWithName:(NSString *)theName {
    [playlist storeWithName:theName];
}

// TGPlaylistDelegate method
- (NSDictionary *)songDataForSongID:(id<SongIDProtocol>)songID {
    
    // Get a mutable copy so we can add a few bits of data.
    NSMutableDictionary *songData = [[_songPoolAPI songDataForSongID:songID] mutableCopy];
    
    // Get the song duration and floor it before adding it to the playlist data (it doesn't really use it)
    double durationDoubleSecs =[[_songPoolAPI songDurationForSongID:songID] doubleValue];
    
    [songData addEntriesFromDictionary:@{@"Duration": [NSNumber numberWithInteger:durationDoubleSecs],
                                         @"SongURL": [_songPoolAPI songURLForSongID:songID],
                                         }];
    return songData;
}

// NSTableViewDataSource delegate methods
-(NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [playlist songsInPlaylist];
}


- (NSView *)tableView:(NSTableView *)tableView
   viewForTableColumn:(NSTableColumn *)tableColumn
                  row:(NSInteger)row {
    
    // Get an existing cell with the MyView identifier if it exists
    TGPlaylistCellView *resultCell = [tableView makeViewWithIdentifier:@"SongCell" owner:self];
    
    NSDictionary *songData = [_songPoolAPI songDataForSongID:[playlist songIDAtIndex:row]];
    
    // Construct the string for the playlist entry.
//    resultCell.layer.backgroundColor = (__bridge CGColorRef)([NSColor whiteColor]);
    resultCell.TitleLabel.stringValue = [songData valueForKey:@"Title"];
    resultCell.AlbumLabel.stringValue = [songData valueForKey:@"Album"];
    resultCell.ArtistLabel.stringValue = [songData valueForKey:@"Artist"];
    
    return resultCell;
}

// NSTableViewDelegate methods


-(void)tableViewSelectionDidChange:(NSNotification *)notification {
    // Find out what row was selected.
    NSInteger selectedRow = [_playlistTableView selectedRow];
   
    // Set it if it is a valid position.
    if (selectedRow >=0) {
        [playlist setPosInPlaylist:selectedRow];
        id<SongIDProtocol> newSongID = [playlist songIDAtIndex:selectedRow];
//        [_songPoolAPI requestSongPlayback:newSongID withStartTimeInSeconds:[NSNumber numberWithFloat:0] makeSweetSpot:NO];
        [_songPoolAPI requestSongPlayback:newSongID withStartTimeInSeconds:nil];
    }

    [_playlistTableView deselectRow:selectedRow];
    // Make the main view controller the first responder.
    [[[self view] window] makeFirstResponder:(NSResponder*)_delegate];
}

@end

File: source/TGPlaylistViewController.swift
//
//  TGPlaylistController.swift
//  reDiscover
//
//  Created by teo on 28/01/16.
//  Copyright © 2016 Teo Sartori. All rights reserved.
//

import Cocoa

protocol PlaylistViewControllerDelegate {
    func getSong(_ songId : SongId) -> TGSong?
    func requestPlayback(_ songId: SongId, startTimeInSeconds: NSNumber)
    func selectIndirectly(_ songId : SongId)
}

public class TGPlaylistViewController : NSViewController, NSTableViewDataSource, NSTableViewDelegate {
    
    @IBOutlet weak var playlistTableView: NSTableView!
    @IBOutlet weak var playlistProgress: NSProgressIndicator!

    
    fileprivate var playlist : TGPlaylist!
    
    var delegate : PlaylistViewControllerDelegate?
    
    public override func awakeFromNib() {
        if playlist == nil {
            playlist = TGPlaylist()
            
            playlist.delegate = self
            playlistTableView.delegate = self
            playlistTableView.dataSource = self
        }
    }
    
    override init?(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    public required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    func storePlaylist(withName fileName : String) {
        playlist.store(withName: fileName)
    }
    
    func addToPlaylist(songWithId songId : SongId) {
        playlist.addSong(withId: songId, atIndex: 0)
        /// TODO: Consider reloading only the changed row
        playlistTableView.reloadData()
    }
    
    func removeFromPlaylist(songWithId songId : SongId) {
        playlist.removeSong(songId)
        playlistTableView.reloadData()
    }
    
    func getNextSongIdToPlay() -> SongId? {
        return playlist.getNextSongIdToPlay()
    }
}

/// TGPlaylistDelegate methods ?
//extension TGPlaylistViewController {
//    
//    func songData(forSongId songId : SongId) -> NSDictionary {
//        
//    }
//}

/// NSTableViewDataSource delegate methods
extension TGPlaylistViewController {

    @objc(numberOfRowsInTableView:) public func numberOfRows(in tableView: NSTableView) -> Int {
        return playlist.songsInPlaylist()
    }
    
    @objc(tableView:viewForTableColumn:row:) public func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        
        guard let resultCell  = tableView.make(withIdentifier: "SongCell", owner: self) as? TGPlaylistCellView,
            let id = playlist.getSongId(atIndex: row),
            let song = delegate?.getSong(id) else {
                return nil
        }
        
        let songData = song.metadataDict()
        guard songData.count > 0 else { return nil }
        
        resultCell.titleLabel.stringValue   = songData["Title"] as! String
        resultCell.albumLabel.stringValue   = songData["Album"] as! String
        resultCell.artistLabel.stringValue  = songData["Artist"] as! String
        
        return resultCell
    }
}

/// NSTableViewDelegate methods
extension TGPlaylistViewController {
    
    public func tableViewSelectionDidChange(_ notification: Notification) {
        let selectedRow = playlistTableView.selectedRow
        if selectedRow >= 0 {
            playlist.positionInPlaylist = selectedRow
            
            if let songId = playlist.getSongId(atIndex: selectedRow) {
                //delegate?.requestPlayback(songId, startTimeInSeconds: 0)
                delegate?.selectIndirectly(songId)
            }
            playlistTableView.deselectRow(selectedRow)
            /// FIXME: Not sure the responder should be self. Used to be _delegate
            self.view.window?.makeFirstResponder(self)
        }
    }
    
    /// Ensure the row height is sufficient.
    public func tableView(_ tableView: NSTableView,heightOfRow row: Int) -> CGFloat {
        return 67
    }
}

extension TGPlaylistViewController : PlaylistDelegate {
    func getSong(_ songId : SongId) -> TGSong? {
        return delegate?.getSong(songId)
    }
}
